### DISCUSSION TITLE
MySQL: How to get the generated primary key when inserting

### DISCUSSION THREAD

**Discussion by ilsubyeega** [diesel-rs/diesel#3365] [Q&A] [answered]
I am testing a disel through mysql and I have a question when trying with examples. \
[I've heard from JDBC that if you add data(row) to a table which has primary key, you can get the generated primary key.](https://stackoverflow.com/a/1915197) Since MySQL does not currently support the RETURNING syntax(however mariadb already has), the example code appears to get the last id of the table after creation. 
https://github.com/diesel-rs/diesel/blob/0622b7d02d7f5ac98a34696250ab3e88fdff0d23/examples/mysql/getting_started_step_3/src/lib.rs#L23-L28
I wonder if it is possible to get the generated key value instead of this way. Thank you.

**Comment by weiznich** [comment, marked as answer]
As the underlying database system does not support a build in way to receive the generated key with one query that's just not possible to implement. Libraries like JDBC just do the two queries internally, which we decided not to do as it hides potential problems. 
Now there are two ways to get the generated key: Using the query from the example or defining `LAST_INSERT_ID()` via `sql_function!` and use that in a separate query. Both require a separate query and depending on your use case you might want to wrap them into a transaction.

Regarding to the `RETURNING` support in mariadb: Supporting that would require a distinct backend implementation for mariadb. Contributions are welcome there, but we as core team don't have a concrete plan to implement that soon.

**Reply by ilsubyeega** [reply]
Oh, Never thought JDBC would be do that internally. I will just use `sql_function!` at this moment.
